### PR TITLE
Chemical requests console

### PR DESCRIPTION
**Pull request by TobleroneSwordfish**
[FEATURE]
## About the PR 
Adds a chemical request console for placing orders with chemistry. One is found outside science and one in medbay on every map. (I had to move/remove a couple of things to make them fit, actual mappers feel free to tell me if it looks wrong)
Might be good to give botany one as well, lmk what you think.

Request interface:
image image

There is a corresponding console in chemistry itself for fulfilling or denying requests, and a PDA alert is sent to all science PDAs when a request is placed.
image
image

Both consoles currently use the same sprite ala QM consoles and because I am not creative: 
image

## Why's this needed? 
Chemistry in theory is able to quickly make and distribute chemicals for the crew via the mail chutes, however this rarely happens as radio and PDA messages tend to be missed and chemistry is often isolated from the rest of the station. (kudos to donut 3, kondaru and clarion for having chemistry front desks)
Adding a formal request system should make it a lot easier for people to request what they want, and hopefully encourage chemistry to fulfil those requests at some point.
The chemistry terminal keeps a history of all requests made and whether or not they were denied, allowing for some investigation into who is to blame for the clown getting access to thermite.

## Changelog

```changelog
(u)LeahTheTech
(*)Added chemical request consoles for requesting chemicals. Find them outside science and in medbay, with another terminal in chemistry to let scientists decide whether to fulfil or deny the clown's request for 200u of space lube.
```

### CODE REVIEW

**Comment by TobleroneSwordfish** [comment]
so somehow the bot managed to make an 8k diff out of adding three machines to cog 2?
image
image

**Comment by ZeWaka** [comment]
You didn't include the firedoor change. Reset and re-do your changes.
image

**Comment by TobleroneSwordfish** [comment]
> You didn't include the firedoor change to cog2. Reset and re-do your changes.
Hmm, I already did that once, let me try again with the second upstream merge

**Comment by TobleroneSwordfish** [comment]
Okay it seems like the issue was caused by #9299 not running the map merge drivers, and has only cropped up here since this is the next PR to modify the same map. The merge seems to be completing fine, so it should be able to be merged with this PR or fixed by a different PR/commit.
Either way, this PR should be fine to review, just ignore the massive diff number.

**Comment by cheekybrdy** [comment]
So sorry about the late notice, seeing as we talked about this on the discord but I forgot an addition and it finally got back to me, maybe put chemcompiler requests as an option, eg. pills, bottle etc. Once again sorry I should have had the opportunity to suggest this earlier if I didn't forget!

**Comment by TobleroneSwordfish** [comment]
I think that'd probably be better just using the notes function since there are a lot of different combinations of pill/bottle/autoinjector sizes that would be a mess to put on a UI. All you have to do currently is put "15u pills please" in notes.
Also now is still a perfectly valid time to make suggestions, that's part of what PRs are for.

**Comment by TobleroneSwordfish** [comment]
oh no, do I dare do it
!merge_upstream

**Comment by TobleroneSwordfish** [comment]
!merge_upstream

**Comment by TobleroneSwordfish** [comment]
map merge conflict is dead! 🦀

File: tgui/packages/tgui/interfaces/ChemRequestReceiver.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    area,
+    state,
+    interactable,
+    age,
+  } = props;
+  // some mild colour hackery to make the text visible on (hopefully) any reagent colour
+  const color_string = "rgba(" + reagent_color[0] + "," + reagent_color[1] + ", " + reagent_color[2] + ", 1)";
+  const lightness = reagent_color.reduce((a, b) => a + b, 0)/3;
+  return (
+    <Section height>
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/ChemRequestReceiver.js:32]: `height` is not a boolean prop, but a number prob. This can probably be removed.

File: tgui/packages/tgui/interfaces/ChemRequestReceiver.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            <Stack.Item>{name} requested</Stack.Item>
+            <Stack.Item align="center"><Box width={16} textAlign="center" backgroundColor={color_string} color={lightness > 255/2 ? "black" : "white"}>{capitalize(reagent_name)} ({volume}u)</Box></Stack.Item>
+            <Stack.Item>from {area} {age} ago. <br /> {notes && `Notes: ${notes}`}</Stack.Item>
+          </Stack>
+        </Flex.Item>
+        <Flex.Item>
+          <Box>
+            {state === "pending" && (
+              <>
+                <Button disabled={!interactable} align="center" width="49.5%" color="red" icon="ban" onClick={() => { act("deny", { id: id }); }}>Deny</Button>
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/ChemRequestReceiver.js:45]: Use `Stack` or `Flex` to put these buttons horizontally next to each other

**Review by Luxizzle — CHANGES_REQUESTED** [review]
First pass while getting a look at this while in game. Not a fan of the search list in the requester window. Long list, you have to scroll down until you finally get to an unlabeled input field and a number field. These should be labeled.

Also, the search list of chemicals should be its own scrollable element so that the inputs are always visible instead of hidden until scrolled down. Some extra padding here would also be nice.

See example:
image

Also, not sure on dev's opinion, but doesn't feel right to me that this includes secret chemicals, this could also list stuff potentially purposefully hidden in the secret repo.

Overall the design of the window is a lot different compared to already existing interfaces, see dispenser for example which it will be living next to a lot: 
image

Small list of things to take note here:

- Use `<Section>`'s to split up parts of your interface. You can label these with a title to even more emphasize their importance and use.
  - Sections can be scrollable by adding `scrollable` to their list of props, use this to create separate scrollable elements.
- We already have a design for chemicals in a UI in the dispenser which displays the color of the chemical in a neat little circle. This design creates a very busy and sometimes unreadable list of items.
- Personal opinion, but I don't like the centering of the items, feels like a lot of empty space around items.
- Use `<Stack>`'s built in padding, don't add custom paddings between them as it will look very out-of-place next to other interfaces.

**Comment by Luxizzle** [comment]
Overall great start :+1: just needs a bit of polish. If you ever need some help, You can always tag me on discord and I can help you out if I have the time.

DM side all looks fine, but someone smarter than me can probably get a better look at that.

**Comment by frawhst** [comment]
merge driver take the wheel

File: tgui/packages/tgui/interfaces/ChemRequestReceiver.js (hunk 2 above)
Comment (TobleroneSwordfish) [reply tgui/packages/tgui/interfaces/ChemRequestReceiver.js:45]: But they're fixed width, not sure how a stack would help?

**Comment by TobleroneSwordfish** [comment]
!merge_upstream

**Comment by TobleroneSwordfish** [comment]
also the map merge seems to have gone badly and added several thousand lines from other PRs again

**Comment by Tarmunora** [comment]
dnm until I add a `var/secret` on `chemical_reaction` and sprinkle it around in secrets. There's some stuff probably shouldn't be showing up on the console,

**Comment by Luxizzle** [comment]
> also the map merge seems to have gone badly and added several thousand lines from other PRs again

Maybe remove them from the maps for now, and they can be added in another pull request later.

**Comment by Luxizzle** [comment]
Small thing, but receiver has no scrollbar 

image

**Comment by Luxizzle** [comment]
Also I suggest using LabeledList for the amount and notes field on the requester.

File: code/obj/machinery/computer/chem_requester.dm [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+		.["selected_reagent"] = src.request.reagent_name
+		.["notes"] = src.request.note
+		.["volume"] = src.request.volume
+
+	ui_static_data(mob/user)
+		. = list()
+		var/list/chems = list()
+		for (var/id in chem_reactions_by_id)
+			var/datum/chemical_reaction/reaction = chem_reactions_by_id[id]
+			if (reaction.result)
```
Comment (Tarmunora) [inline code/obj/machinery/computer/chem_requester.dm:43]:
```suggestion
			if (reaction.result && !reaction.hidden)
```

**Comment by TobleroneSwordfish** [comment]
Remade without map changes here: #9607
